openedx/openedx-platform
# feat: course_roles setup

## Description

This pull request adds the new CourseRoles model that will be used for creating roles with associated permissions that can be assigned to users on a course, as an org-wide course role (the user has access as if they have the role on all courses in the org), or instance-wide course role (the user has access as if they have the role on all courses within the instance). 

This initial PR does not create any roles, but does setup the ability to do so in the future.

Additionally this PR adds code at each location where a user's django_commentclientrole or student_courseaccessrole is currently checked. At each of these locations an or clause was added to check if the user has the permission for this level of access. Which permission to check was determined based on which permissions the role would have (if migrated to the new course_roles system) and which of those permissions best corresponds to the functionality of the code.

If implemented correctly this code will have no immediate impact on system users, but will instead set the code up for future iterations that will allow implementation of new roles using this system.

## Supporting information

This PR is the initial work for the [Tech Spec related to Course Roles](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3824648277/RBAC+Tech+Spec). 

## Testing instructions

In order to run the code fully, a waffle flag (course_roles.use_permission_checks) should be created. If the waffle flag is not present or is false, the permissions checks will return false. Only if the waffle flag is true will the permissions checks check the db for access.

As noted previously, there is no intended change in functionality for a user. Testing should reflect no noticeable user experience changes as there are currently no course_roles_roles created and no users who are assigned the new course_roles_roles.

## Other information

Requested Areas of Additional Focus
- code related to adding data for xblocks
- default value of the waffle flag (if it is not created)

Starting Point for PR Review
- [openedx/core/djangoapps/course_roles/docs/decisions/0001-course_level_roles.rst](https://github.com/openedx/edx-platform/pull/33609/files/eab726e8a1267b4c0e31cbab0df6a7df9196c9bd#diff-09d45b3a47f61c47edabbca9774601e4de1887ef0b621084af693cd06ce62003) provides context on these changes and may be a good starting point for reviewing the PR
